sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -501,6 +501,10 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
         case LessThanOrEqual(l, r) => GreaterThan(l, r)
         // not(not(e))  =>  e
         case Not(e) => e
+        // De Morgan's law: !(a || b) => !a && !b
```

Comment:
I think it's a boolean simplification because we can only do further optimisation when `Not` is on leaf node, the above `Not(LessThanOrEqual)` or `Not(Not)` is a good example.